### PR TITLE
Tidy up namespace.

### LINKED ISSUE #1570: Tidy up external namespace
Currently we're being very messy about what gets exported into the external namespace.

I've taken a :shrug: approach to this up to now, assuming that people will read the documentation and not delve into undocumented features. However, this might not be such a good idea any more, as IDEs and things like [ConquerOfCompletions](https://github.com/neoclide/coc.nvim) (one of my rewards for shipping 1.0 is going to be setting this up, finally) scan the namespace, making stuff we don't document much more discoverable.

So, should we lock down the namespace to just the stuff we support, going forward? 

Thoughts @benjeffery, @molpopgen, @petrelharp, @grahamgower (and anyone else)?

### PR DESCRIPTION
**Pull request by jeromekelleher**
- Only export stuff we're documenting
- Add __all__ to be more polite
- Hide away the deprecation facade stuff to avoid cluttering up the
namespace.

Closes #1570

Haven't done the underscores for private methods thing. It's pretty tedious...

### CODE REVIEW

File: msprime/__init__.py [hunk 1]
```diff
@@ -17,97 +18,216 @@
...
+from msprime.mutations import InfiniteAllelesMutationModel
+from msprime.mutations import JC69MutationModel
+from msprime.mutations import MatrixMutationModel
+from msprime.mutations import MutationModel
+from msprime.mutations import PAMMutationModel
+from msprime.mutations import sim_mutations
+from msprime.mutations import SLiMMutationModel
 
-from types import ModuleType
+# We have to include the
```
Comment (petrelharp) [inline msprime/__init__.py:61]: include the what?

File: tests/test_demography.py [hunk 2]
```diff
@@ -6140,7 +6154,7 @@ def test_struct_metadata(self):
 
 class TestLineageMovementEvents:
     def test_as_lineage_movement_abstract(self):
-        e = msprime.LineageMovementEvent(1234)
+        e = msprime.demography.LineageMovementEvent(1234)
```
Comment (benjeffery) [inline tests/test_demography.py:6157]: I'm not following why the others get `demog_mod` and this one doesn't.

File: tests/test_deprecation.py [hunk 3]
```diff
@@ -68,10 +68,26 @@ def test_moved_to_tskit(self):
             "REVERSE",
         ]:
             with pytest.warns(FutureWarning):
-                getattr(msprime, name)
+                msp_version = getattr(msprime, name)
+            assert msp_version is getattr(tskit, name)
```
Comment (benjeffery) [inline tests/test_deprecation.py:72]: Nice.

**Review by benjeffery — APPROVED** [review]
LGTM, one question.

File: tests/test_demography.py (hunk 2 above)
Comment (jeromekelleher) [reply tests/test_demography.py:6157]: No good reason, I'll change it.

**Pull request merged**